zebrad/src/components/inbound.rs

```diff
@@ -369,66 +392,97 @@ impl Service<zn::Request> for Inbound {
-                // unless the request is empty.
+                // unless the request is empty, or it reaches a response limit.
                 if hashes.is_empty() {
                     return async { Ok(zn::Response::Nil) }.boxed();
                 }
 
-                // Correctness:
-                //
-                // We can't use `call_all` here, because it can hold one buffer slot per concurrent
-                // future, until the `CallAll` struct is dropped. We can't hold those slots in this
-                // future because:
-                // * we're not sure when the returned future will complete, and
-                // * we don't limit how many returned futures can be concurrently running
-                // https://github.com/tower-rs/tower/blob/master/tower/src/util/call_all/common.rs#L112
-                use futures::stream::TryStreamExt;
-                hashes
-                    .iter()
-                    .cloned()
-                    .map(|hash| zs::Request::Block(hash.into()))
-                    .map(|request| state.clone().oneshot(request))
-                    .collect::<futures::stream::FuturesOrdered<_>>()
-                    .try_filter_map(|response| async move {
-                        Ok(match response {
-                            zs::Response::Block(Some(block)) => Some(block),
-                            zs::Response::Block(None) => None,
+                let state = state.clone();
+
+                async move {
+                    let mut blocks: Vec<InventoryResponse<Arc<Block>, block::Hash>> = Vec::new();
+                    let mut total_size = 0;
+
+                    // Ignore any block hashes past the response limit.
+                    // This saves us expensive database lookups.
+                    for &hash in hashes.iter().take(GETDATA_MAX_BLOCK_COUNT) {
+                        // We check the limit after including at least one block, so that we can
+                        // send blocks greater than 1 MB (but only one at a time)
+                        if total_size >= GETDATA_SENT_BYTES_LIMIT {
+                            break;
+                        }
+
+                        let response = state.clone().ready().await?.call(zs::Request::Block(hash.into())).await?;
+
+                        // Add the block responses to the list, while updating the size limit.
+                        //
+                        // If there was a database error, return the error,
+                        // and stop processing further chunks.
+                        match response {
+                            zs::Response::Block(Some(block)) => {
+                                // If checking the serialized size of the block performs badly,
+                                // return the size from the state using a wrapper type.
+                                total_size += block.zcash_serialized_size();
+
+                                blocks.push(Available(block))
```

Comment:
Could we check that `blocks.is_empty()` before adding blocks past the limit here:

```suggestion
                                let new_total_size = total_size + block.zcash_serialized_size();

                                if new_total_size <= GETDATA_SENT_BYTES_LIMIT || blocks.is_empty() {
                                    total_size = new_total_size;
                                    blocks.push(Available(block));
                                };
```

Reply:
We've already done the work to retrieve this block from the database, so we might as well send it on the network.

The limit isn't a network protocol limit, it's just a performance/denial of service limit.